### PR TITLE
Week2hardware

### PR DESCRIPTION
**Pull request by barulicm**
Opening for review purposes. Will not be merged until #55 is merged into master.

### CODE REVIEW

File: hardware_applications/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+   ```cmake ..```
+   
+4. Build all demos
+
+   ```make```
+
+## How to Rebuild the STSL ##
+
+Binaries and headers for the STSL are provided in the STSL folder. These should work for most circumstances. If you experience problems building the demos with the provided binaries, you may need to rebuild the STSL from source.
```
Comment (ebarnette) [inline hardware_applications/README.md]: Update to include explanation of STSL like in week1

File: hardware_applications/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+   
+3. Configure the project with CMake
+
+   ```cmake ..```
+   
+4. Build the STSL
+
+   ```make```
+   
+5. Copy the binaries back to the hardware_applications project.
```
Comment (ebarnette) [inline hardware_applications/README.md]: Update file paths like in week1

File: hardware_applications/STSL/include/STSL/LinuxUtils.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#ifndef TRAININGSUPPORTLIBRARY_LINUXUTILS_H
+#define TRAININGSUPPORTLIBRARY_LINUXUTILS_H
+
+#include "OSUtils.h"
+#include <vector>
+
+class LinuxUtils : public OSUtils {
+public:
+    virtual std::string FindRobot() override;
```
Comment (ebarnette) [inline hardware_applications/STSL/include/STSL/LinuxUtils.h]: use namespaces?

File: hardware_applications/STSL/include/STSL/OSUtils.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#ifndef TRAININGSUPPORTLIBRARY_OSUTILS_H
+#define TRAININGSUPPORTLIBRARY_OSUTILS_H
+
+#include <string>
+#include <chrono>
+
+class OSUtils {
+public:
+    virtual std::string FindRobot() = 0;
```
Comment (ebarnette) [inline hardware_applications/STSL/include/STSL/OSUtils.h]: use namespaces?

File: hardware_applications/STSL/include/STSL/RJRobot.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    uint8_t LightValue();
+
+    void SetFloodlight(bool on);
+
+    void SetMotor(const MotorPort &port, const int &speed);
+
+    void StopMotors();
+
+    void Wait(std::chrono::microseconds duration);
```
Comment (ebarnette) [inline hardware_applications/STSL/include/STSL/RJRobot.h]: namespaces?

File: hardware_applications/week2/final_helpers.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    do {
+        executeOneCommand(robot, commands.substr(currentPosition, nextSpacePosition));
+        currentPosition = nextSpacePosition + 1;
+        nextSpacePosition = commands.find(' ', currentPosition);
+    } while(nextSpacePosition != std::string::npos);
+}
+
+void executeOneCommand(RJRobot &robot, const std::string &command) {
+
+    if(command == "forward") {
```
Comment (ebarnette) [inline hardware_applications/week2/final_helpers.cpp:18]: This might be a good place to use cases instead of a ton of if/elseif?

File: hardware_applications/week2/final_helpers.cpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    else if(command == "waitForButton") {
+        while (!robot.IsButtonPressed()) {
+            robot.Wait(50ms);
+        }
+    }
+    else if(command == "stop") {
+        robot.StopMotors();
+    }
+    else if(command == "dance") {
```
Comment (ebarnette) [inline hardware_applications/week2/final_helpers.cpp:46]: instead of ??? can you put something more descriptive?

File: hardware_applications/week2/final_helpers.cpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        // ???
+    }
+    else {
+        std::cerr << "Unkown command: " << command << std::endl;
```
Comment (ebarnette) [inline hardware_applications/week2/final_helpers.cpp:50]: namespaces?

File: hardware_applications/week2/starter_main.cpp [hunk 9]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (ebarnette) [inline hardware_applications/week2/starter_main.cpp:1]: do we want to give them includes?

**Review by ebarnette — DISMISSED** [review]
Mostly good overall. A small number of tweaks.

File: hardware_applications/week2/final_helpers.cpp (hunk 6 above)
Comment (barulicm) [reply hardware_applications/week2/final_helpers.cpp:18]: It might... except...
https://stackoverflow.com/questions/650162/why-switch-statement-cannot-be-applied-on-strings
TL;DR you can't switch on a string

File: hardware_applications/week2/final_helpers.cpp (hunk 7 above)
Comment (barulicm) [reply hardware_applications/week2/final_helpers.cpp:46]: Eh, the point of this block is they can make it do whatever they want.

File: hardware_applications/week2/starter_main.cpp (hunk 9 above)
Comment (barulicm) [reply hardware_applications/week2/starter_main.cpp:1]: Up to you. IIRC we're teaching includes this week, so it seemed like a good idea to make them type them out themselves to reinforce the learning they just did.

**Comment by barulicm** [comment]
@jgkamat I've switched to using CMake's ExternalProject mechanism to automatically pull and build the latest STSL. It seems pretty magical. 
It looks like it's working on my machine and @sastorer 's. If you're more familiar with ExternalProject, can you glance at it and make sure I'm not breaking any best practices with that tool?

**Review by ebarnette — DISMISSED** [review]

**Comment by jgkamat** [comment]
Minus #74, I think this is good. Keep in mind having something 'pull from head' with an ExternalProjectAdd will force people to clean rebuild to get the latest dependencies (which is fixed by specifying a commit to vendor from), but I think that will be fine for now. We can always vendor newer commits later.

**Review by ebarnette — APPROVED** [review]
